File: MathCore.DSP/Filters/ChebyshevFilter.cs
using System.Runtime.Serialization;

using static System.Math;
// ReSharper disable InconsistentNaming

namespace MathCore.DSP.Filters;

/// <summary>Фильтр Чебышева</summary>
[KnownType(typeof(ChebyshevLowPass))]
[KnownType(typeof(ChebyshevHighPass))]
[KnownType(typeof(ChebyshevBandPass))]
[KnownType(typeof(ChebyshevBandStop))]
public abstract class ChebyshevFilter : AnalogBasedFilter
{
    /// <summary>Вычисляет арксинус гиперболический</summary>
    /// <param name="x">Аргумент функции</param>
    /// <returns>Значение арксинуса гиперболического</returns>
    protected static double arcsh(double x) => Log(x + Sqrt(x * x + 1));

    /// <summary>Вычисляет арккосинус гиперболический</summary>
    /// <param name="x">Аргумент функции</param>
    /// <returns>Значение арккосинуса гиперболического</returns>
    protected static double arcch(double x) => Log(x + Sqrt(x * x - 1));

    /// <summary>Возвращает нормированные полюса фильтра Чебышева первого рода</summary>
    /// <param name="N">Порядок фильтра</param>
    /// <param name="EpsP">Неоднородность АЧХ в полосе пропускания</param>
    /// <param name="W0">Нормирующий множитель</param>
    /// <returns>Перечисление комплексных полюсов</returns>
    protected static IEnumerable<Complex> GetNormedPolesI(int N, double EpsP, double W0 = 1)
    {
        var beta = arcsh(1 / EpsP) / N;
        var sh = Sinh(beta) * W0;
        var ch = Cosh(beta) * W0;
        if (N.IsOdd()) yield return -sh;    // Если порядок фильтра нечётный, то первым добавляем центральный полюс
        var r = N % 2;                      // Нечётность порядка фильтра
        for (var (i, dth) = (r, Consts.pi05 / N); i < N; i += 2)   // Расчёт полюсов
        {
            var th = dth * (i - r + 1);

            var sin = Sin(th);
            var cos = Cos(th);
            yield return new(-sh * sin, +ch * cos);
            yield return new(-sh * sin, -ch * cos);
        }
    }

    /// <summary>Возвращает нормированные нули и полюса фильтра Чебышева второго рода</summary>
    /// <param name="N">Порядок фильтра</param>
    /// <param name="EpsS">Неоднородность АЧХ в полосе заграждения</param>
    /// <param name="W0">Нормирующий множитель</param>
    /// <returns>Кортеж массивов нулей и полюсов</returns>
    protected static (Complex[] Zeros, Complex[] Poles) GetNormedPolesII(int N, double EpsS, double W0 = 1)
    {
        var beta = arcsh(EpsS) / N;
        var sh = Sinh(beta);
        var ch = Cosh(beta);

        var r = N % 2;                              // Нечётность порядка фильтра
        var zeros = new Complex[N - r];
        var poles = new Complex[N];                 // Массив полюсов фильтра
        if (N.IsOdd())
            poles[0] = -W0 / sh;                    // Если порядок фильтра нечётный, то первым добавляем центральный полюс
        for (var (i, dth) = (r, Consts.pi05 / N); i < poles.Length; i += 2)   // Расчёт полюсов
        {
            var th = dth * (i - r + 1);
            var (sin, cos) = Complex.SinCos(th);
            var z = new Complex(-sh * sin, ch * cos);
            var norm = W0 / z.Power;
            (poles[i], poles[i + 1]) = Complex.Conjugate(z.Re * norm, z.Im * norm);
            (zeros[i - r], zeros[i - r + 1]) = Complex.Conjugate(0, W0 / Cos(th));
        }

        return (zeros, poles);
    }

    /// <summary>Тип фильтра</summary>
    public ChebyshevType FilterType { get; }

    /// <inheritdoc />
    /// <param name="Type">Тип фильтра I или II</param>
    protected ChebyshevFilter(double[] B, double[] A, Specification Spec, ChebyshevType Type) : base(B, A, Spec) => FilterType = Type;
}

File: MathCore.DSP/Filters/Python/filters/chebyshev_filter.py
# Базовый класс фильтра Чебышева (порт C# ChebyshevFilter + ChebyshevType)
from enum import Enum

import numpy as np

from .analog_based_filter import AnalogBasedFilter, Specification


class ChebyshevType(Enum):
    """Тип (род) фильтра Чебышева."""
    I = 1             # равноволновая пульсация в полосе пропускания
    II = 2            # равноволновая пульсация в полосе заграждения
    IICorrected = 3   # II рода с точным попаданием в частоты среза и заграждения


class ChebyshevFilter(AnalogBasedFilter):
    """
    Базовый класс фильтра Чебышева.

    Содержит вычисление нормированных нулей и полюсов для обоих родов.
    """

    def __init__(self, B: np.ndarray, A: np.ndarray, spec: Specification,
                 filter_type: ChebyshevType):
        super().__init__(B, A, spec)
        self.filter_type = filter_type

    # ── Гиперболические обратные функции ──────────────────────────────────────

    @staticmethod
    def arcsh(x: float) -> float:
        """Арксинус гиперболический arcsh(x) = ln(x + √(x²+1))."""
        return np.log(x + np.sqrt(x * x + 1.0))

    @staticmethod
    def arcch(x: float) -> float:
        """Аркосинус гиперболический arcch(x) = ln(x + √(x²-1))."""
        return np.log(x + np.sqrt(x * x - 1.0))

    # ── Полюса Чебышева I рода ────────────────────────────────────────────────

    @staticmethod
    def get_normed_poles_i(N: int, EpsP: float, W0: float = 1.0) -> list:
        """
        Нормированные полюса ФНЧ Чебышева I рода.

        Полюса размещаются на эллипсе с полуосями sh (горизонтальная)
        и ch (вертикальная).

        Параметры
        ---------
        N    : порядок фильтра
        EpsP : неравномерность АЧХ в полосе пропускания
        W0   : нормирующий множитель частоты
        """
        beta = ChebyshevFilter.arcsh(1.0 / EpsP) / N
        sh = np.sinh(beta) * W0   # горизонтальная полуось эллипса
        ch = np.cosh(beta) * W0   # вертикальная полуось эллипса

        r = N % 2  # нечётность порядка
        poles = []
        if N % 2 != 0:
            poles.append(complex(-sh, 0.0))  # центральный вещественный полюс
        dth = np.pi / 2.0 / N
        for i in range(r, N, 2):
            th = dth * (i - r + 1)
            sin_th, cos_th = np.sin(th), np.cos(th)
            poles.append(complex(-sh * sin_th, +ch * cos_th))
            poles.append(complex(-sh * sin_th, -ch * cos_th))
        return poles

    # ── Полюса и нули Чебышева II рода ────────────────────────────────────────

    @staticmethod
    def get_normed_poles_ii(N: int, EpsS: float, W0: float = 1.0):
        """
        Нормированные нули и полюса ФНЧ Чебышева II рода.

        Нули расположены на мнимой оси (вне полосы пропускания),
        полюса — на деформированном эллипсе.

        Параметры
        ---------
        N    : порядок фильтра
        EpsS : неравномерность АЧХ в полосе заграждения
        W0   : нормирующий множитель частоты

        Возвращает
        ----------
        (zeros, poles) : кортеж комплексных массивов
        """
        beta = ChebyshevFilter.arcsh(EpsS) / N
        sh = np.sinh(beta)
        ch = np.cosh(beta)

        r = N % 2
        zeros = []
        poles = []

        if N % 2 != 0:
            poles.append(complex(-W0 / sh, 0.0))  # вещественный полюс нечётного порядка

        dth = np.pi / 2.0 / N
        for i in range(r, N, 2):
            th = dth * (i - r + 1)
            sin_th, cos_th = np.sin(th), np.cos(th)

            # промежуточный комплексный вектор
            z_re = -sh * sin_th
            z_im = ch * cos_th
            z_pow2 = z_re ** 2 + z_im ** 2  # |z|²
            norm = W0 / z_pow2

            # полюса: W0 / z.conjugate() = W0 * z.re / |z|² + j * W0 * (-z.im) / |z|²
            poles.append(complex(z_re * norm, +z_im * norm))
            poles.append(complex(z_re * norm, -z_im * norm))

            # нули на мнимой оси: ±j·W0/cos(th)
            zeros.append(complex(0.0, +W0 / cos_th))
            zeros.append(complex(0.0, -W0 / cos_th))

        return zeros, poles
